# **CA684 Machine Learning Assignment**

>[CA684 Machine Learning Assignment](#scrollTo=x-FI6VANcFtC)

>[Introduction](#scrollTo=X4pm68VEBVzw)

>[References used :](#scrollTo=_tFQo-bXC1gL)

>[Import packages](#scrollTo=Lm8xWe8JvoNv)

>[Reading Data](#scrollTo=ntlbSBkcv81p)

>[Basic EDA](#scrollTo=hPmojeEDzk7X)

>[Top Category With Bert](#scrollTo=n1kpcRgKwoG-)

>>[Data Preparing and preprocessing](#scrollTo=g3MXInkcP9XZ)

>>[Model prep and training](#scrollTo=p_RpR8pqVTEt)

>>[Evaluation and Results](#scrollTo=mMy8ow3AAzq1)

>[Color With Bert](#scrollTo=CsyPu1SDOnM7)

>>[Data Preparing and preprocessing](#scrollTo=xSxugDrfOnM8)

>>[Model prep and training](#scrollTo=Tn9I2ew5OnM9)

>>[Evaluation and Results](#scrollTo=9oMtbrSCBP0D)

>[Bottom Category With Bert](#scrollTo=JlnQkKfodwNj)

>>[Data Preparing and preprocessing](#scrollTo=EZvJmzPzdwNk)

>>[Model prep and training](#scrollTo=Gm78L05-dwNw)

>>[Evaluation and Results](#scrollTo=gsUtb1pPBBXU)

>[Bottom Category With Random Forest](#scrollTo=pDzxCnW1dwN8)

>>[Preprocessing](#scrollTo=0nD7RFd6ue5c)

>>[Training](#scrollTo=9xE3JoXvCpc7)

>>[Results](#scrollTo=OQcuTiZMwWGC)

>[Top Category With Random Forest](#scrollTo=xks2PRs4wcNS)

>>[Preprocessing](#scrollTo=SZi3gYeIwcNY)

>>[Training](#scrollTo=TMSB0Cx_wcNe)

>>[Results](#scrollTo=z6zjvkWIwcNh)

>[Color With Random Forest](#scrollTo=_2Tz_KLyzH-l)

>>[Preprocessing](#scrollTo=WukpKntUzH-n)

>>[Training](#scrollTo=dAj8281-zH-s)

>>[Results](#scrollTo=mnvOJEd3zH-u)

>[Submission](#scrollTo=riQrC1c428sp)

>[What can I do to improve this?](#scrollTo=Gz_gC8ACIsZ-)



# Introduction

Hello and welcome to my notebook. You can see the contents of the notebook just above this.
The main aim of this notebook is to use the data provided by Etsy for multiclass classfication. I have initially tried several different methods but I am only showcases random forest and multimodal learning with Bert in this notebook.
I have referred several websites including stackoverflow, articles, papers and even github repo's in my attempt to compelete this assignemnt.
I have made sure to include all the references in this notebook and in my report.


To solve this problem, I have put myself in the perspective of the transformer model and asked, "***Which values from the dataset can be used to make predictions for each of the columns?***"



Disclaimer: My python skills are not the greatest...(yet)
So I have a lot of repeated code but I hope the indexing above helps you nagivate easily.

## **This version of the project is intended for public viewing. To adhere to the NDA agreement with Etsy, certain elements may have been removed or omitted**

#  Credits







*   I have referred to the GitHub repository at https://github.com/georgian-io/Multimodal-Toolkit **This is the github which contains the multimodal package based on the research paper available at https://aclanthology.org/2021.maiworkshop-1.10.pdf.**
* The documentation for the above was also really helpful to understand what was going on. https://multimodal-toolkit.readthedocs.io/en/latest/index.html
*   This notebook also refers the [example](https://colab.research.google.com/github/huggingface/blog/blob/master/notebooks/trainer/01_text_classification.ipynb#scrollTo=bwl3I_VGAZXb) from HuggingFace for text classificaion on the GLUE dataset.
* Apart from these I have of course used/referred snippets of code from websites like https://scikit-learn.org and https://stackoverflow.com/. I have tried to give credits where ever I have used them in this notebook



<a id="1"></a> <br>
# Import packages

In [ ]:
!pip install multimodal-transformers



Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached pandas-1.3.5-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.5 MB)
  Using cached numpy-1.21.6-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.2
    Uninstalling numpy-1.24.2:
      Successfully uninstalled numpy-1.24.2
  Attempting uninstall: pandas
    Found existing installation: pandas 2.0.0
    Uninstalling pandas-2.0.0:
      Successfully uninstalled pandas-2.0.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.12.0 requires numpy<1.24,>=1.22, but you have numpy 1.21.6 which is incompatible.
google-colab 1.0.0 requires pandas~=1.5.3, but you have pandas 1.3.5 which is incompatible.
arviz 0.15.1 requires scipy>=1.8.0, but you ha

In [ ]:
pip install --force-reinstall pandas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached pandas-2.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
  Using cached pytz-2023.3-py2.py3-none-any.whl (502 kB)
  Using cached numpy-1.24.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
  Using cached tzdata-2023.3-py2.py3-none-any.whl (341 kB)
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
  Attempting uninstall: pytz
    Found existing installation: pytz 2023.3
    Uninstalling pytz-2023.3:
      Successfully uninstalled pytz-2023.3
  Attempting uninstall: tzdata
    Found existing installation: tzdata 2023.3
    Uninstalling tzdata-2023.3:
      Successfully uninstalled tzdata-2023.3
  Attempting uninstall: six
    Found existing installation: six 1.16.0
    Uninstalling six-1.16.0:
      Successfully uninstalled six-1.16.0
  Attempting uninstall: nump

In [ ]:
from dataclasses import dataclass, field
import json
import logging
import os
from typing import Optional

import numpy as np
import pandas as pd
from transformers import (
    AutoTokenizer,
    AutoConfig,
    Trainer,
    EvalPrediction,
    set_seed
)
from transformers.training_args import TrainingArguments

from multimodal_transformers.data import load_data_from_folder
from multimodal_transformers.model import TabularConfig
from multimodal_transformers.model import AutoModelWithTabular

logging.basicConfig(level=logging.INFO)
os.environ['COMET_MODE'] = 'DISABLED'


/usr/local/lib/python3.9/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


Around 30% of the time, transformers import fails but this is fixed by simply restarting kernal.

In [ ]:
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from sklearn.model_selection import train_test_split
from PIL import Image


<a id="2"></a> <br>
# Reading Data

In [ ]:
def get_parquet_data(folder):
    """ Read parquet data from mutiple files """


    dataframes = [
        pd.read_parquet(os.path.join(folder, filename))
        for filename in list(os.listdir(folder))
    ]

    return pd.concat(dataframes)

In [ ]:
from google.colab import drive

drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
data_train = get_parquet_data('/content/gdrive/MyDrive/parquet/train')
data_test = get_parquet_data('/content/gdrive/MyDrive/parquet/test')

<a id="2"></a> <br>
# Basic EDA

In [ ]:
f"Number of rows in train: {len(data_train):,}"
f"Number of unique products in train: {len(data_train['product_id'].unique()):,}"
f"Number of columns in train: {len(data_train.columns):,}"
list(data_train.columns)

In [ ]:
data_train['color_id'].value_counts()

color_id
1     32537
17    31194
2     26972
4     22891
9     15675
11    15499
14    15243
16    14892
0     12664
7     10728
19     9920
18     7406
12     6639
5      6337
10     5346
13     5226
3      2922
6      2237
15     1156
8         1
Name: count, dtype: int64

In [ ]:
data_train['color_text'].value_counts()


color_text
black        32537
white        31194
blue         26972
brown        22891
green        15675
pink         15499
red          15243
silver       14892
beige        12664
gold         10728
other         9920
yellow        7406
purple        6639
clear         6337
orange        5346
rainbow       5226
bronze        2922
copper        2237
rose gold     1156
grey             1
Name: count, dtype: int64

In [ ]:
pd.value_counts(data_train['top_category_text'], sort=True, ascending=False)

In [ ]:
pd.value_counts(data_train['top_category_id'], sort=True, ascending=False)

top_category_id
8     53200
6     42348
5     32028
13    14427
0     12610
3     12426
10    10811
12    10355
14    10333
9      9457
1      8741
7      8177
4      7378
2      6645
11     6549
Name: count, dtype: int64

In [ ]:
figsize = (15, 6)
column = 'top_category_text'
x_limit = 60000
x_step = 10000
color = 'darkcyan'

figure, ax = plt.subplots(figsize=figsize)
pd.value_counts(
    data_train[column],
    sort=True, ascending=False).plot.barh(color=color)
plt.title(f"Number of Products per Category")
xlabels = [f'{x // 1000:,}K' for x in range(0, x_limit, x_step)]
plt.xticks(range(0, x_limit, x_step), xlabels)
plt.xlabel('Number of products')
plt.setp(ax.get_xticklabels()[0], visible=False)
plt.show()

In [ ]:
f"Number of bottom categories in train: {len(data_train['bottom_category_text'].unique()):,}"

'Number of bottom categories in train: 2,782'

In [ ]:
pd.value_counts(data_train['bottom_category_id'], sort=True, ascending=False)

bottom_category_id
511      98
2203     98
6064     98
1652     97
6267     97
         ..
6807     44
1481     43
6546     42
2342     42
11158    41
Name: count, Length: 2782, dtype: int64

I will be using three different model for the three labels that need to be predicted. We will start with top category, color and then finally bottom category

#Top Category With Bert

## Data Preparing and preprocessing

There isnt much preprocessing and prepping other than just cleaning and arranging the data because of several errors that I faced during my several trial runs.

While doing my testing and inital trial, I Used a subset of the data. This helped me run multiple ideas on free colab and get faster results. Of course the accuracy and F1 scores were bad but at least I got an idea that helped me sort out what to do with the whole dataset. The codes below are commented because I didnt use them while running the entire dataset. They were used for my inital trials.

In [ ]:
# import random
# data_train = data_train.sample(n=20000,random_state=3)

In [ ]:
# data_test = data_test.sample(n=2000,random_state=3)

The below code is for splitting the train dataset into training and validation.
For this first model we are going to predict 'top_category_id'

In [ ]:
from sklearn.model_selection import train_test_split

df_training, data_val = train_test_split(
    data_train,
    test_size=0.1,
    stratify=data_train['top_category_id'],
    random_state=42)

The codes below are for fixing some indexes and ensuring there are no duplicate indexes because that messes up when inputing into the model.
I faced these errors and fixed them by debugging and referring stackoverflow

In [ ]:
df_training.reset_index(drop=True, inplace=True)
data_test.reset_index(drop=True, inplace=True)
data_val.reset_index(drop=True, inplace=True)

In [ ]:
df_training["Unnamed: 0"] = list(range(df_training.shape[0]))
data_val["Unnamed: 0"] = list(range(df_training.shape[0], df_training.shape[0] + data_val.shape[0]))
data_test["Unnamed: 0"] = list(range(df_training.shape[0] + data_val.shape[0], df_training.shape[0] + data_val.shape[0] + data_test.shape[0]))

Storing as csv for the dataloader

In [ ]:
df_training.to_csv('train.csv',index=False)
data_test.to_csv('test.csv',index=False)
data_val.to_csv('val.csv',index=False)

## Model prep and training
We use Data Classes to hold each of our arguments for the model, data, and training.

In [ ]:
@dataclass
class ModelArguments:
  """
  Arguments pertaining to which model/config/tokenizer we are going to fine-tune from.
  """

  model_name_or_path: str = field(
      metadata={"help": "Path to pretrained model or model identifier from huggingface.co/models"}
  )
  config_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained config name or path if not the same as model_name"}
  )
  tokenizer_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
  )
  cache_dir: Optional[str] = field(
      default=None, metadata={"help": "Where do you want to store the pretrained models downloaded from s3"}
  )


@dataclass
class MultimodalDataTrainingArguments:
  """
  Arguments pertaining to how we combine tabular features
  Using `HfArgumentParser` we can turn this class
  into argparse arguments to be able to specify them on
  the command line.
  """

  data_path: str = field(metadata={
                            'help': 'the path to the csv file containing the dataset'
                        })
  column_info_path: str = field(
      default=None,
      metadata={
          'help': 'the path to the json file detailing which columns are text, categorical, numerical, and the label'
  })

  column_info: dict = field(
      default=None,
      metadata={
          'help': 'a dict referencing the text, categorical, numerical, and label columns'
                  'its keys are text_cols, num_cols, cat_cols, and label_col'
  })

  categorical_encode_type: str = field(default='ohe',
                                        metadata={
                                            'help': 'sklearn encoder to use for categorical data',
                                            'choices': ['ohe', 'binary', 'label', 'none']
                                        })
  numerical_transformer_method: str = field(default='yeo_johnson',
                                            metadata={
                                                'help': 'sklearn numerical transformer to preprocess numerical data',
                                                'choices': ['yeo_johnson', 'box_cox', 'quantile_normal', 'none']
                                            })
  task: str = field(default="classification",
                    metadata={
                        "help": "The downstream training task",
                        "choices": ["classification", "regression"]
                    })

  mlp_division: int = field(default=4,
                            metadata={
                                'help': 'the ratio of the number of '
                                        'hidden dims in a current layer to the next MLP layer'
                            })
  combine_feat_method: str = field(default='individual_mlps_on_cat_and_numerical_feats_then_concat',
                                    metadata={
                                        'help': 'method to combine categorical and numerical features, '
                                                'see README for all the method'
                                    })
  mlp_dropout: float = field(default=0.1,
                              metadata={
                                'help': 'dropout ratio used for MLP layers'
                              })
  numerical_bn: bool = field(default=True,
                              metadata={
                                  'help': 'whether to use batchnorm on numerical features'
                              })
  use_simple_classifier: str = field(default=True,
                                      metadata={
                                          'help': 'whether to use single layer or MLP as final classifier'
                                      })
  mlp_act: str = field(default='relu',
                        metadata={
                            'help': 'the activation function to use for finetuning layers',
                            'choices': ['relu', 'prelu', 'sigmoid', 'tanh', 'linear']
                        })
  gating_beta: float = field(default=0.2,
                              metadata={
                                  'help': "the beta hyperparameters used for gating tabular data "
                                          "see https://www.aclweb.org/anthology/2020.acl-main.214.pdf"
                              })

  def __post_init__(self):
      assert self.column_info != self.column_info_path
      if self.column_info is None and self.column_info_path:
          with open(self.column_info_path, 'r') as f:
              self.column_info = json.load(f)

For model we can specify any supported HuggingFace model classes (see README for more details) as well as any AutoModel that are from the supported model classes. For this assignment I am using **BERT** For the data specifications, we need to specify a dictionary that specifies which columns are the `text` columns, `numerical feature` columns, `categorical feature` column, and the `label` column. If we are doing classification, we can also specify what each of the labels means in the label column through the `label list`. We can also specifiy these columns using a path to a json file with the argument `column_info_path` to `MultimodalDataTrainingArguments`.

In [ ]:
text_cols = ['title', 'description', 'tags'] # these are the columns that I am using to predict top_category_id
cat_cols = ['type'] #tabular
numerical_cols = []

column_info_dict = {
    'text_cols': text_cols,
    'num_cols': numerical_cols,
    'cat_cols': cat_cols,
    'label_col': 'top_category_id', # what i need to predict
    'label_list': data_train['top_category_id'].unique() #the different labels that the product can fall into aka the classes
}


model_args = ModelArguments(
    model_name_or_path='bert-base-uncased'
)

data_args = MultimodalDataTrainingArguments(
    data_path='.',
    combine_feat_method='gating_on_cat_and_num_feats_then_sum', #?
    column_info=column_info_dict,
    task='classification',

)

training_args = TrainingArguments(
    output_dir="./logs/model_name",
    logging_dir="./logs/runs",
    overwrite_output_dir=True,
    do_train=True,
    do_eval=True,
    per_device_train_batch_size=32,
    num_train_epochs=1,
    evaluation_strategy='steps',
    eval_steps=300,
    logging_steps=300
)

set_seed(training_args.seed)

***HuggingFace tokenizer***

In [ ]:
tokenizer_path_or_name = model_args.tokenizer_name if model_args.tokenizer_name else model_args.model_name_or_path
print('Specified tokenizer: ', tokenizer_path_or_name)
tokenizer = AutoTokenizer.from_pretrained(
    tokenizer_path_or_name,
    cache_dir=model_args.cache_dir,
)

Specified tokenizer:  bert-base-uncased


The function `load_data_from_folder` expects a path to a folder that contains `train.csv`, `test.csv`, and/or `val.csv` containing the respective split datasets.

In [ ]:
# Get Datasets
train_dataset, val_dataset, test_dataset = load_data_from_folder(
    data_args.data_path,
    data_args.column_info['text_cols'],
    tokenizer,
    label_col=data_args.column_info['label_col'],
    label_list=data_args.column_info['label_list'],
    categorical_cols=data_args.column_info['cat_cols'],
    numerical_cols=data_args.column_info['num_cols'],
    sep_text_token_str=tokenizer.sep_token,
    numerical_transformer_method = 'none'
)

In [ ]:
num_labels = len(np.unique(train_dataset.labels))
num_labels

15

In [ ]:
config = AutoConfig.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        cache_dir=model_args.cache_dir,
    )
tabular_config = TabularConfig(num_labels=num_labels,
                               cat_feat_dim=train_dataset.cat_feats.shape[1],
                               numerical_feat_dim=0,
                               **vars(data_args))
config.tabular_config = tabular_config

In [ ]:
import torch
torch.cuda.is_available()

True

The below code is to make sure we use the GPU, Although I am not entirely sure if it needs to be there if you are running on colab. If you are using jupiter notebook then definitely yes but make sure CUDA and CNN are also installed.

In [ ]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)


cuda:0


In [ ]:
model = AutoModelWithTabular.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        config=config,
        cache_dir=model_args.cache_dir
    ).to(device)

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertWithTabular: ['cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.bias']
- This IS expected if you are initializing BertWithTabular from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertWithTabular from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertWithTabular were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['tabular_

In [ ]:
import numpy as np
from scipy.special import softmax
from sklearn.metrics import (
    auc,
    precision_recall_curve,
    roc_auc_score,
    f1_score,
    confusion_matrix,
    matthews_corrcoef,
)

def calc_classification_metrics(p: EvalPrediction):
  predictions = p.predictions[0]
  pred_labels = np.argmax(predictions, axis=1)
  pred_scores = softmax(predictions, axis=1)[:, 1]
  labels = p.label_ids
  if len(np.unique(labels)) == 2:  # binary classification
      roc_auc_pred_score = roc_auc_score(labels, pred_scores)
      precisions, recalls, thresholds = precision_recall_curve(labels,
                                                                pred_scores)
      fscore = (2 * precisions * recalls) / (precisions + recalls)
      fscore[np.isnan(fscore)] = 0
      ix = np.argmax(fscore)
      threshold = thresholds[ix].item()
      pr_auc = auc(recalls, precisions)
      tn, fp, fn, tp = confusion_matrix(labels, pred_labels, labels=[0, 1]).ravel()
      result = {'roc_auc': roc_auc_pred_score,
                'threshold': threshold,
                'pr_auc': pr_auc,
                'recall': recalls[ix].item(),
                'precision': precisions[ix].item(), 'f1': fscore[ix].item(),
                'tn': tn.item(), 'fp': fp.item(), 'fn': fn.item(), 'tp': tp.item()
                }
  else:
      acc = (pred_labels == labels).mean()
      f1 = f1_score(y_true=labels, y_pred=pred_labels,average='micro')
      result = {
          "acc": acc,
          "f1": f1,
          "acc_and_f1": (acc + f1) / 2,
          "mcc": matthews_corrcoef(labels, pred_labels)
      }

  return result

In [ ]:
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=calc_classification_metrics,
)

In [ ]:
%%time
trainer.train()

/usr/local/lib/python3.9/dist-packages/transformers/optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


CPU times: user 2h 34min 36s, sys: 33.3 s, total: 2h 35min 9s
Wall time: 2h 34min 42s


TrainOutput(global_step=6905, training_loss=0.5104286144126765, metrics={'train_runtime': 9282.3974, 'train_samples_per_second': 23.802, 'train_steps_per_second': 0.744, 'total_flos': 5.854406827064525e+16, 'train_loss': 0.5104286144126765, 'epoch': 1.0})

In [ ]:
trainer.save_model("./topcategorymodel")

## Evaluation and Results

In [ ]:
val_predictions = trainer.predict(val_dataset)

In [ ]:
from sklearn.metrics import classification_report

Lets check our F1 score on the validation dataset

In [ ]:
print(classification_report(data_val['top_category_id'],val_predictions.predictions[0].argmax(axis=1)))

              precision    recall  f1-score   support

           0       0.83      0.83      0.83      1261
           1       0.76      0.70      0.73       874
           2       0.85      0.88      0.87       664
           3       0.89      0.90      0.90      1243
           4       0.91      0.91      0.91       738
           5       0.93      0.97      0.95      3203
           6       0.90      0.89      0.89      4235
           7       0.89      0.91      0.90       818
           8       0.89      0.92      0.90      5320
           9       0.87      0.90      0.89       946
          10       0.82      0.85      0.84      1081
          11       0.95      0.92      0.94       655
          12       0.95      0.96      0.96      1035
          13       0.89      0.86      0.88      1443
          14       0.75      0.58      0.66      1033

    accuracy                           0.89     24549
   macro avg       0.87      0.87      0.87     24549
weighted avg       0.88   

In [ ]:
%cp -av ./topcategorymodel /content/gdrive/MyDrive/etsy_trained_models/
#code to move the model to google drive so that I dont lose it when i restart runtime

'./topcategorymodel' -> '/content/gdrive/MyDrive/etsy_trained_models/topcategorymodel'
'./topcategorymodel/config.json' -> '/content/gdrive/MyDrive/etsy_trained_models/topcategorymodel/config.json'
'./topcategorymodel/pytorch_model.bin' -> '/content/gdrive/MyDrive/etsy_trained_models/topcategorymodel/pytorch_model.bin'
'./topcategorymodel/training_args.bin' -> '/content/gdrive/MyDrive/etsy_trained_models/topcategorymodel/training_args.bin'


In [ ]:
test_predictions = trainer.predict(test_dataset)

RuntimeError: ignored

Below code is to load model from local(or drive)





Most of the time I run out of memmory or my runtime crashes or CUDA gets blocked when I get to my test predictions. In that case I save the model and simply load it again and run the test predictions.

In [ ]:
model = AutoModelWithTabular.from_pretrained("/content/gdrive/MyDrive/etsy_trained_models/topcategorymodel/")
model.to(device)
model.eval()

BertWithTabular(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_a

In [ ]:
val_dataset

In [ ]:
test_dataset

In [ ]:
test_dataloader = torch.utils.data.DataLoader(test_dataset, batch_size=32)
test_predictions = []
with torch.no_grad():
    for batch in test_dataloader:
        # Move the input data to the device where the model is located
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        cat_feats = batch['cat_feats'].to(device)

        # Make predictions using the model
        outputs = model(input_ids, attention_mask, cat_feats=cat_feats)[1]

        # Convert the model outputs to a numpy array
        numpy_outputs = outputs.argmax(axis=1).cpu().numpy().tolist()

        # Add the predictions to the list of predictions
        test_predictions.extend(numpy_outputs)

Storing the results

In [ ]:
data_test['top_category_id'] = test_predictions

Saving the predictions and product_id to predictions.csv
I will add the predictions for each model into this csv and then covert it to parquet for final submission

In [ ]:
data_test[['product_id','top_category_id']].to_csv('/content/gdrive/MyDrive/prediction.csv')

# Color With Bert

## Data Preparing and preprocessing

There isnt much preprocessing and prepping other than just cleaning and arranging the data because of several errors that I faced during my several trial runs.

In [ ]:
data_train['color_text'].value_counts()


color_text
black        32537
white        31194
blue         26972
brown        22891
green        15675
pink         15499
red          15243
silver       14892
beige        12664
gold         10728
other         9920
yellow        7406
purple        6639
clear         6337
orange        5346
rainbow       5226
bronze        2922
copper        2237
rose gold     1156
grey             1
Name: count, dtype: int64

There is only one row for Grey. It will be impossible to train this when we split training and validation grey will be there only in one of them... For this reason I am dropping grey.

In [ ]:
data_train = data_train[data_train['color_id'] != 8]

In [ ]:
color_trainingID, _ = pd.factorize(data_train['color_text'])

This model of bert that I use needs continous variables so I am making my own mapping. I could use color_id but we just deleted grey which would make it not continous.

In [ ]:
data_train['color_trainingID'] = color_trainingID

In [ ]:
data_train['color_trainingID'].value_counts()


color_trainingID
1     32537
7     31194
2     26972
3     22891
5     15675
11    15499
6     15243
9     14892
13    12664
12    10728
14     9920
4      7406
0      6639
8      6337
10     5346
16     5226
18     2922
15     2237
17     1156
Name: count, dtype: int64

While doing my testing and inital trial, I Used a subset of the data. This helped me run multiple ideas on free colab and get faster results. Of course the accuracy and F1 scores were bad but at least I got an idea that helped me sort out what to do with the whole dataset. The codes below are commented because I didnt use them while running the entire dataset. They were used for my inital trials.

In [ ]:
# import random
# data_train = data_train.sample(n=20000,random_state=3)

In [ ]:
# data_test = data_test.sample(n=2000,random_state=3)

The below code is for splitting the train dataset into training and validation.
For this  model we are going to predict 'color_trainingID' ( later we will map it to color_id)

In [ ]:
from sklearn.model_selection import train_test_split

df_training, data_val = train_test_split(
    data_train,
    test_size=0.1,
    stratify=data_train['color_trainingID'],
    random_state=42)

The codes below are for fixing some indexes and ensuring there are no duplicate indexes because that messes up when inputing into the model.
I faced these errors and fixed them by debugging and referring stackoverflow

In [ ]:
df_training.reset_index(drop=True, inplace=True)
data_test.reset_index(drop=True, inplace=True)
data_val.reset_index(drop=True, inplace=True)

In [ ]:
df_training["Unnamed: 0"] = list(range(df_training.shape[0]))
data_val["Unnamed: 0"] = list(range(df_training.shape[0], df_training.shape[0] + data_val.shape[0]))
data_test["Unnamed: 0"] = list(range(df_training.shape[0] + data_val.shape[0], df_training.shape[0] + data_val.shape[0] + data_test.shape[0]))

Storing it as csv because the dataset loader function takes csv

In [ ]:
df_training.to_csv('train.csv',index=False)
data_test.to_csv('test.csv',index=False)
data_val.to_csv('val.csv',index=False)

## Model prep and training

In [ ]:
@dataclass
class ModelArguments:
  """
  Arguments pertaining to which model/config/tokenizer we are going to fine-tune from.
  """

  model_name_or_path: str = field(
      metadata={"help": "Path to pretrained model or model identifier from huggingface.co/models"}
  )
  config_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained config name or path if not the same as model_name"}
  )
  tokenizer_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
  )
  cache_dir: Optional[str] = field(
      default=None, metadata={"help": "Where do you want to store the pretrained models downloaded from s3"}
  )


@dataclass
class MultimodalDataTrainingArguments:
  """
  Arguments pertaining to how we combine tabular features
  Using `HfArgumentParser` we can turn this class
  into argparse arguments to be able to specify them on
  the command line.
  """

  data_path: str = field(metadata={
                            'help': 'the path to the csv file containing the dataset'
                        })
  column_info_path: str = field(
      default=None,
      metadata={
          'help': 'the path to the json file detailing which columns are text, categorical, numerical, and the label'
  })

  column_info: dict = field(
      default=None,
      metadata={
          'help': 'a dict referencing the text, categorical, numerical, and label columns'
                  'its keys are text_cols, num_cols, cat_cols, and label_col'
  })

  categorical_encode_type: str = field(default='ohe',
                                        metadata={
                                            'help': 'sklearn encoder to use for categorical data',
                                            'choices': ['ohe', 'binary', 'label', 'none']
                                        })
  numerical_transformer_method: str = field(default='yeo_johnson',
                                            metadata={
                                                'help': 'sklearn numerical transformer to preprocess numerical data',
                                                'choices': ['yeo_johnson', 'box_cox', 'quantile_normal', 'none']
                                            })
  task: str = field(default="classification",
                    metadata={
                        "help": "The downstream training task",
                        "choices": ["classification", "regression"]
                    })

  mlp_division: int = field(default=4,
                            metadata={
                                'help': 'the ratio of the number of '
                                        'hidden dims in a current layer to the next MLP layer'
                            })
  combine_feat_method: str = field(default='individual_mlps_on_cat_and_numerical_feats_then_concat',
                                    metadata={
                                        'help': 'method to combine categorical and numerical features, '
                                                'see README for all the method'
                                    })
  mlp_dropout: float = field(default=0.1,
                              metadata={
                                'help': 'dropout ratio used for MLP layers'
                              })
  numerical_bn: bool = field(default=True,
                              metadata={
                                  'help': 'whether to use batchnorm on numerical features'
                              })
  use_simple_classifier: str = field(default=True,
                                      metadata={
                                          'help': 'whether to use single layer or MLP as final classifier'
                                      })
  mlp_act: str = field(default='relu',
                        metadata={
                            'help': 'the activation function to use for finetuning layers',
                            'choices': ['relu', 'prelu', 'sigmoid', 'tanh', 'linear']
                        })
  gating_beta: float = field(default=0.2,
                              metadata={
                                  'help': "the beta hyperparameters used for gating tabular data "
                                          "see https://www.aclweb.org/anthology/2020.acl-main.214.pdf"
                              })

  def __post_init__(self):
      assert self.column_info != self.column_info_path
      if self.column_info is None and self.column_info_path:
          with open(self.column_info_path, 'r') as f:
              self.column_info = json.load(f)

In [ ]:
text_cols = ['title', 'description', 'tags'] # these are the columns that I am using to predict top_category_id
cat_cols = ['type']
numerical_cols = []

column_info_dict = {
    'text_cols': text_cols,
    'num_cols': numerical_cols,
    'cat_cols': cat_cols,
    'label_col': 'color_trainingID', # what i need to predict
    'label_list': data_train['color_trainingID'].unique() #the different labels that the product can fall into aka the classes
}


model_args = ModelArguments(
    model_name_or_path='bert-base-uncased'
)

data_args = MultimodalDataTrainingArguments(
    data_path='.',
    combine_feat_method='gating_on_cat_and_num_feats_then_sum', #?
    column_info=column_info_dict,
    task='classification',

)

training_args = TrainingArguments(
    output_dir="./logs/model_name",
    logging_dir="./logs/runs",
    overwrite_output_dir=True,
    do_train=True,
    do_eval=True,
    per_device_train_batch_size=32,
    num_train_epochs=1,
    evaluation_strategy='steps',
    eval_steps=700,
    logging_steps=700
)

set_seed(training_args.seed)

In [ ]:
tokenizer_path_or_name = model_args.tokenizer_name if model_args.tokenizer_name else model_args.model_name_or_path
print('Specified tokenizer: ', tokenizer_path_or_name)
tokenizer = AutoTokenizer.from_pretrained(
    tokenizer_path_or_name,
    cache_dir=model_args.cache_dir,
)

Specified tokenizer:  bert-base-uncased


In [ ]:
# Get Datasets
train_dataset, val_dataset, test_dataset = load_data_from_folder(
    data_args.data_path,
    data_args.column_info['text_cols'],
    tokenizer,
    label_col=data_args.column_info['label_col'],
    label_list=data_args.column_info['label_list'],
    categorical_cols=data_args.column_info['cat_cols'],
    numerical_cols=data_args.column_info['num_cols'],
    sep_text_token_str=tokenizer.sep_token,
    numerical_transformer_method = 'none'
)

In [ ]:
num_labels = len(np.unique(train_dataset.labels))
num_labels

19

In [ ]:
config = AutoConfig.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        cache_dir=model_args.cache_dir,
    )
tabular_config = TabularConfig(num_labels=num_labels,
                               cat_feat_dim=train_dataset.cat_feats.shape[1],
                               numerical_feat_dim=0,
                               **vars(data_args))
config.tabular_config = tabular_config

In [ ]:
import torch
torch.cuda.is_available()

True

The below code is to make sure we use the GPU, Although I am not entirely sure if it needs to be there if you are running on colab. If you are using jupiter notebook then definitely yes but make sure CUDA and CNN are also installed.

In [ ]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)


cuda:0


In [ ]:
model = AutoModelWithTabular.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        config=config,
        cache_dir=model_args.cache_dir
    ).to(device)

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertWithTabular: ['cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.predictions.bias']
- This IS expected if you are initializing BertWithTabular from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertWithTabular from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertWithTabular were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['tabular_

In [ ]:
import numpy as np
from scipy.special import softmax
from sklearn.metrics import (
    auc,
    precision_recall_curve,
    roc_auc_score,
    f1_score,
    confusion_matrix,
    matthews_corrcoef,
)

def calc_classification_metrics(p: EvalPrediction):
  predictions = p.predictions[0]
  pred_labels = np.argmax(predictions, axis=1)
  pred_scores = softmax(predictions, axis=1)[:, 1]
  labels = p.label_ids
  if len(np.unique(labels)) == 2:  # binary classification
      roc_auc_pred_score = roc_auc_score(labels, pred_scores)
      precisions, recalls, thresholds = precision_recall_curve(labels,
                                                                pred_scores)
      fscore = (2 * precisions * recalls) / (precisions + recalls)
      fscore[np.isnan(fscore)] = 0
      ix = np.argmax(fscore)
      threshold = thresholds[ix].item()
      pr_auc = auc(recalls, precisions)
      tn, fp, fn, tp = confusion_matrix(labels, pred_labels, labels=[0, 1]).ravel()
      result = {'roc_auc': roc_auc_pred_score,
                'threshold': threshold,
                'pr_auc': pr_auc,
                'recall': recalls[ix].item(),
                'precision': precisions[ix].item(), 'f1': fscore[ix].item(),
                'tn': tn.item(), 'fp': fp.item(), 'fn': fn.item(), 'tp': tp.item()
                }
  else:
      acc = (pred_labels == labels).mean()
      f1 = f1_score(y_true=labels, y_pred=pred_labels,average='micro')
      result = {
          "acc": acc,
          "f1": f1,
          "acc_and_f1": (acc + f1) / 2,
          "mcc": matthews_corrcoef(labels, pred_labels)
      }

  return result

In [ ]:
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=calc_classification_metrics,
)

In [ ]:
%%time
trainer.train()

/usr/local/lib/python3.9/dist-packages/transformers/optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


CPU times: user 1h 49min 35s, sys: 27.6 s, total: 1h 50min 3s
Wall time: 1h 49min 30s


TrainOutput(global_step=6905, training_loss=1.4664747720175597, metrics={'train_runtime': 6570.1804, 'train_samples_per_second': 33.627, 'train_steps_per_second': 1.051, 'total_flos': 5.85458910077184e+16, 'train_loss': 1.4664747720175597, 'epoch': 1.0})

In [ ]:
trainer.save_model("./colormodel")

## Evaluation and Results

In [ ]:
val_predictions = trainer.predict(val_dataset)

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(data_val['color_trainingID'],val_predictions.predictions[0].argmax(axis=1)))

              precision    recall  f1-score   support

           0       0.71      0.57      0.63       664
           1       0.59      0.69      0.64      3254
           2       0.57      0.68      0.62      2697
           3       0.59      0.69      0.64      2289
           4       0.66      0.48      0.56       741
           5       0.68      0.58      0.63      1567
           6       0.60      0.58      0.59      1524
           7       0.53      0.59      0.56      3119
           8       0.65      0.62      0.64       634
           9       0.72      0.78      0.75      1489
          10       0.66      0.50      0.57       535
          11       0.65      0.59      0.62      1550
          12       0.66      0.65      0.66      1073
          13       0.52      0.35      0.42      1266
          14       0.64      0.46      0.53       992
          15       0.72      0.68      0.70       224
          16       0.46      0.38      0.42       523
          17       0.69    

In [ ]:
test_predictions = trainer.predict(test_dataset) # here my test predictions is getting CUDA error
#So I will save the model and use it again below.

RuntimeError: ignored

In [ ]:
%cp -av ./colormodel /content/gdrive/MyDrive/etsy_trained_models/
#code to move the model to google drive so that I dont lose it when i restart runtime

'./colormodel' -> '/content/gdrive/MyDrive/etsy_trained_models/colormodel'
'./colormodel/config.json' -> '/content/gdrive/MyDrive/etsy_trained_models/colormodel/config.json'
'./colormodel/pytorch_model.bin' -> '/content/gdrive/MyDrive/etsy_trained_models/colormodel/pytorch_model.bin'
'./colormodel/training_args.bin' -> '/content/gdrive/MyDrive/etsy_trained_models/colormodel/training_args.bin'


Below code is to load model from local(or drive)





Most of the time I run out of memmory or my runtime crashes or CUDA gets blocked when I get to my test predictions. In that case I save the model and simply load it again and run the test predictions.

In [ ]:
model = AutoModelWithTabular.from_pretrained("/content/gdrive/MyDrive/etsy_trained_models/colormodel/")
model.to(device)
model.eval()

BertWithTabular(
  (bert): BertModel(
    (embeddings): BertEmbeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (token_type_embeddings): Embedding(2, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): BertEncoder(
      (layer): ModuleList(
        (0-11): 12 x BertLayer(
          (attention): BertAttention(
            (self): BertSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): BertSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_a

In [ ]:
test_dataloader = torch.utils.data.DataLoader(test_dataset, batch_size=32)
test_predictions = []
with torch.no_grad():
    for batch in test_dataloader:
        # Move the input data to the device where the model is located
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        cat_feats = batch['cat_feats'].to(device)

        # Make predictions using the model
        outputs = model(input_ids, attention_mask, cat_feats=cat_feats)[1]

        # Convert the model outputs to a numpy array
        numpy_outputs = outputs.argmax(axis=1).cpu().numpy().tolist()

        # Add the predictions to the list of predictions
        test_predictions.extend(numpy_outputs)

Since I am using color_trainingID, I need to map the predictions to color_id

In [ ]:
map1 = set(data_train[['color_id','color_trainingID']].to_records(index=False).tolist())

In [ ]:
data_train.head(25)

In [ ]:
mapping = {y:x for (x,y) in map1}

In [ ]:
test_predictions

[10,
 5,
 5,
 5,
 6,
 5,
 5,
 13,
 8,
 7,
 13,
 0,
 7,
 1,
 0,
 6,
 7,
 8,
 8,
 8,
 8,
 3,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 8,
 6,
 7,
 7,
 1,
 1,
 1,
 14,
 7,
 3,
 7,
 7,
 2,
 7,
 2,
 1,
 8,
 9,
 7,
 7,
 15,
 12,
 5,
 9,
 7,
 1,
 1,
 1,
 1,
 14,
 3,
 3,
 6,
 9,
 9,
 9,
 10,
 8,
 8,
 6,
 16,
 16,
 6,
 13,
 16,
 7,
 7,
 6,
 7,
 6,
 3,
 5,
 7,
 1,
 1,
 3,
 6,
 1,
 2,
 13,
 1,
 10,
 2,
 3,
 5,
 12,
 3,
 3,
 7,
 7,
 2,
 7,
 13,
 7,
 3,
 7,
 11,
 7,
 13,
 2,
 7,
 2,
 7,
 7,
 7,
 7,
 3,
 9,
 2,
 3,
 9,
 9,
 7,
 7,
 3,
 15,
 7,
 1,
 7,
 5,
 9,
 7,
 2,
 3,
 5,
 5,
 5,
 2,
 5,
 5,
 3,
 1,
 1,
 14,
 1,
 7,
 6,
 2,
 1,
 2,
 14,
 1,
 3,
 3,
 3,
 1,
 1,
 14,
 3,
 3,
 3,
 1,
 3,
 3,
 1,
 0,
 3,
 6,
 1,
 14,
 1,
 1,
 7,
 16,
 16,
 1,
 1,
 1,
 3,
 4,
 13,
 3,
 13,
 14,
 13,
 10,
 12,
 18,
 6,
 8,
 11,
 14,
 7,
 11,
 7,
 7,
 11,
 14,
 9,
 2,
 18,
 15,
 14,
 7,
 12,
 7,
 7,
 18,
 8,
 3,
 3,
 2,
 17,
 11,
 2,
 7,
 7,
 5,
 1,
 2,
 2,
 0,
 11,
 11,
 3,
 3,
 14,
 2,
 14,
 3,
 14,
 10,
 3,
 2,
 2,
 6,
 14,
 

In [ ]:
test_predictions_mapped = [mapping[x] for x in test_predictions]

In [ ]:
pd.DataFrame(test_predictions_mapped).to_csv('/content/gdrive/MyDrive/colorpredictions.csv')

# Bottom Category With Bert

## Data Preparing and preprocessing

There isnt much preprocessing and prepping other than just cleaning and arranging the data because of several errors that I faced during my several trial runs.

In [ ]:
data_train['bottom_category_text'].value_counts()


In [ ]:
bottom_trainingID, _ = pd.factorize(data_train['bottom_category_text'])

This model of bert that I use needs continous variables so I am making my own mapping. I could use bottom_category_id but it is not continous

In [ ]:
data_train['bottom_trainingID'] = bottom_trainingID

In [ ]:
data_train['bottom_trainingID'].value_counts()


bottom_trainingID
1311    98
796     98
2348    98
1523    97
437     97
        ..
992     44
497     43
554     42
1464    42
2198    41
Name: count, Length: 2782, dtype: int64

While doing my testing and inital trial, I Used a subset of the data. This helped me run multiple ideas on free colab and get faster results. Of course the accuracy and F1 scores were bad but at least I got an idea that helped me sort out what to do with the whole dataset. The codes below are commented because I didnt use them while running the entire dataset. They were used for my inital trials.

In [ ]:
# import random
# data_train = data_train.sample(n=20000,random_state=3)

In [ ]:
# data_test = data_test.sample(n=2000,random_state=3)

The below code is for splitting the train dataset into training and validation.
For this  model we are going to predict 'bottom_trainingID'

In [ ]:
from sklearn.model_selection import train_test_split

df_training, data_val = train_test_split(
    data_train,
    test_size=0.1,
    stratify=data_train['bottom_trainingID'],
    random_state=42)

The codes below are for fixing some indexes and ensuring there are no duplicate indexes because that messes up when inputing into the model.
I faced these errors and fixed them by debugging and referring stackoverflow

In [ ]:
df_training.reset_index(drop=True, inplace=True)
data_test.reset_index(drop=True, inplace=True)
data_val.reset_index(drop=True, inplace=True)

In [ ]:
df_training["Unnamed: 0"] = list(range(df_training.shape[0]))
data_val["Unnamed: 0"] = list(range(df_training.shape[0], df_training.shape[0] + data_val.shape[0]))
data_test["Unnamed: 0"] = list(range(df_training.shape[0] + data_val.shape[0], df_training.shape[0] + data_val.shape[0] + data_test.shape[0]))

Storing it as csv because the dataset loader function takes csv

In [ ]:
df_training.to_csv('train.csv',index=False)
data_test.to_csv('test.csv',index=False)
data_val.to_csv('val.csv',index=False)

## Model prep and training

In [ ]:
@dataclass
class ModelArguments:
  """
  Arguments pertaining to which model/config/tokenizer we are going to fine-tune from.
  """

  model_name_or_path: str = field(
      metadata={"help": "Path to pretrained model or model identifier from huggingface.co/models"}
  )
  config_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained config name or path if not the same as model_name"}
  )
  tokenizer_name: Optional[str] = field(
      default=None, metadata={"help": "Pretrained tokenizer name or path if not the same as model_name"}
  )
  cache_dir: Optional[str] = field(
      default=None, metadata={"help": "Where do you want to store the pretrained models downloaded from s3"}
  )


@dataclass
class MultimodalDataTrainingArguments:
  """
  Arguments pertaining to how we combine tabular features
  Using `HfArgumentParser` we can turn this class
  into argparse arguments to be able to specify them on
  the command line.
  """

  data_path: str = field(metadata={
                            'help': 'the path to the csv file containing the dataset'
                        })
  column_info_path: str = field(
      default=None,
      metadata={
          'help': 'the path to the json file detailing which columns are text, categorical, numerical, and the label'
  })

  column_info: dict = field(
      default=None,
      metadata={
          'help': 'a dict referencing the text, categorical, numerical, and label columns'
                  'its keys are text_cols, num_cols, cat_cols, and label_col'
  })

  categorical_encode_type: str = field(default='ohe',
                                        metadata={
                                            'help': 'sklearn encoder to use for categorical data',
                                            'choices': ['ohe', 'binary', 'label', 'none']
                                        })
  numerical_transformer_method: str = field(default='yeo_johnson',
                                            metadata={
                                                'help': 'sklearn numerical transformer to preprocess numerical data',
                                                'choices': ['yeo_johnson', 'box_cox', 'quantile_normal', 'none']
                                            })
  task: str = field(default="classification",
                    metadata={
                        "help": "The downstream training task",
                        "choices": ["classification", "regression"]
                    })

  mlp_division: int = field(default=4,
                            metadata={
                                'help': 'the ratio of the number of '
                                        'hidden dims in a current layer to the next MLP layer'
                            })
  combine_feat_method: str = field(default='individual_mlps_on_cat_and_numerical_feats_then_concat',
                                    metadata={
                                        'help': 'method to combine categorical and numerical features, '
                                                'see README for all the method'
                                    })
  mlp_dropout: float = field(default=0.1,
                              metadata={
                                'help': 'dropout ratio used for MLP layers'
                              })
  numerical_bn: bool = field(default=True,
                              metadata={
                                  'help': 'whether to use batchnorm on numerical features'
                              })
  use_simple_classifier: str = field(default=True,
                                      metadata={
                                          'help': 'whether to use single layer or MLP as final classifier'
                                      })
  mlp_act: str = field(default='relu',
                        metadata={
                            'help': 'the activation function to use for finetuning layers',
                            'choices': ['relu', 'prelu', 'sigmoid', 'tanh', 'linear']
                        })
  gating_beta: float = field(default=0.2,
                              metadata={
                                  'help': "the beta hyperparameters used for gating tabular data "
                                          "see https://www.aclweb.org/anthology/2020.acl-main.214.pdf"
                              })

  def __post_init__(self):
      assert self.column_info != self.column_info_path
      if self.column_info is None and self.column_info_path:
          with open(self.column_info_path, 'r') as f:
              self.column_info = json.load(f)

In [ ]:
text_cols = ['title', 'description', 'tags'] # these are the columns that I am using to predict top_category_id
cat_cols = ['type']
numerical_cols = []

column_info_dict = {
    'text_cols': text_cols,
    'num_cols': numerical_cols,
    'cat_cols': cat_cols,
    'label_col': 'bottom_trainingID', # what i need to predict
    'label_list': data_train['bottom_trainingID'].unique() #the different labels that the product can fall into aka the classes
}


model_args = ModelArguments(
    model_name_or_path='bert-base-uncased'
)

data_args = MultimodalDataTrainingArguments(
    data_path='.',
    combine_feat_method='gating_on_cat_and_num_feats_then_sum', #?
    column_info=column_info_dict,
    task='classification',

)

training_args = TrainingArguments(
    output_dir="./logs/model_name",
    logging_dir="./logs/runs",
    overwrite_output_dir=True,
    do_train=True,
    do_eval=True,
    per_device_train_batch_size=32,
    num_train_epochs=1,
    evaluation_strategy='steps',
    eval_steps=700,
    logging_steps=700
)

set_seed(training_args.seed)

In [ ]:
tokenizer_path_or_name = model_args.tokenizer_name if model_args.tokenizer_name else model_args.model_name_or_path
print('Specified tokenizer: ', tokenizer_path_or_name)
tokenizer = AutoTokenizer.from_pretrained(
    tokenizer_path_or_name,
    cache_dir=model_args.cache_dir,
)

Specified tokenizer:  bert-base-uncased


In [ ]:
# Get Datasets
train_dataset, val_dataset, test_dataset = load_data_from_folder(
    data_args.data_path,
    data_args.column_info['text_cols'],
    tokenizer,
    label_col=data_args.column_info['label_col'],
    label_list=data_args.column_info['label_list'],
    categorical_cols=data_args.column_info['cat_cols'],
    numerical_cols=data_args.column_info['num_cols'],
    sep_text_token_str=tokenizer.sep_token,
    numerical_transformer_method = 'none'
)

In [ ]:
num_labels = len(np.unique(train_dataset.labels))
num_labels

2782

2782 classes, this is going to be hard xD

In [ ]:
config = AutoConfig.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        cache_dir=model_args.cache_dir,
    )
tabular_config = TabularConfig(num_labels=num_labels,
                               cat_feat_dim=train_dataset.cat_feats.shape[1],
                               numerical_feat_dim=0,
                               **vars(data_args))
config.tabular_config = tabular_config

In [ ]:
import torch
torch.cuda.is_available()

True

The below code is to make sure we use the GPU, Although I am not entirely sure if it needs to be there if you are running on colab. If you are using jupiter notebook then definitely yes but make sure CUDA and CNN are also installed.

In [ ]:
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)


cuda:0


In [ ]:
model = AutoModelWithTabular.from_pretrained(
        model_args.config_name if model_args.config_name else model_args.model_name_or_path,
        config=config,
        cache_dir=model_args.cache_dir
    ).to(device)

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertWithTabular: ['cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight']
- This IS expected if you are initializing BertWithTabular from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertWithTabular from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertWithTabular were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifi

In [ ]:
import numpy as np
from scipy.special import softmax
from sklearn.metrics import (
    auc,
    precision_recall_curve,
    roc_auc_score,
    f1_score,
    confusion_matrix,
    matthews_corrcoef,
)

def calc_classification_metrics(p: EvalPrediction):
  predictions = p.predictions[0]
  pred_labels = np.argmax(predictions, axis=1)
  pred_scores = softmax(predictions, axis=1)[:, 1]
  labels = p.label_ids
  if len(np.unique(labels)) == 2:  # binary classification
      roc_auc_pred_score = roc_auc_score(labels, pred_scores)
      precisions, recalls, thresholds = precision_recall_curve(labels,
                                                                pred_scores)
      fscore = (2 * precisions * recalls) / (precisions + recalls)
      fscore[np.isnan(fscore)] = 0
      ix = np.argmax(fscore)
      threshold = thresholds[ix].item()
      pr_auc = auc(recalls, precisions)
      tn, fp, fn, tp = confusion_matrix(labels, pred_labels, labels=[0, 1]).ravel()
      result = {'roc_auc': roc_auc_pred_score,
                'threshold': threshold,
                'pr_auc': pr_auc,
                'recall': recalls[ix].item(),
                'precision': precisions[ix].item(), 'f1': fscore[ix].item(),
                'tn': tn.item(), 'fp': fp.item(), 'fn': fn.item(), 'tp': tp.item()
                }
  else:
      acc = (pred_labels == labels).mean()
      f1 = f1_score(y_true=labels, y_pred=pred_labels,average='micro')
      result = {
          "acc": acc,
          "f1": f1,
          "acc_and_f1": (acc + f1) / 2,
          "mcc": matthews_corrcoef(labels, pred_labels)
      }

  return result

In [ ]:
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    compute_metrics=calc_classification_metrics,
)

In [ ]:
%%time
trainer.train()

/usr/local/lib/python3.9/dist-packages/transformers/optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


CPU times: user 1h 49min 40s, sys: 38.6 s, total: 1h 50min 18s
Wall time: 1h 49min 52s


TrainOutput(global_step=6905, training_loss=5.614308479193067, metrics={'train_runtime': 6592.617, 'train_samples_per_second': 33.513, 'train_steps_per_second': 1.047, 'total_flos': 5.998825449219686e+16, 'train_loss': 5.614308479193067, 'epoch': 1.0})

In [ ]:
trainer.save_model("./bottomcategorymodel")

## Evaluation and Results

In [ ]:
val_predictions = trainer.predict(val_dataset)

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(data_val['bottom_trainingID'],val_predictions.predictions[0].argmax(axis=1)))

              precision    recall  f1-score   support

           0       0.23      0.78      0.36         9
           1       0.50      0.29      0.36         7
           2       0.00      0.00      0.00         6
           3       0.00      0.00      0.00         9
           4       0.00      0.00      0.00         9
           5       0.00      0.00      0.00         9
           6       0.30      0.89      0.44         9
           7       0.08      0.33      0.12         9
           8       0.00      0.00      0.00         8
           9       0.00      0.00      0.00         9
          10       0.00      0.00      0.00         9
          11       0.00      0.00      0.00         9
          12       0.00      0.00      0.00        10
          13       0.00      0.00      0.00         9
          14       0.00      0.00      0.00         9
          15       0.00      0.00      0.00         6
          16       0.16      0.75      0.26         8
          17       0.62    

Maybe if with more epochs I could get better result but there is always the chance of model overfitting but the reason I didnt run it for more epochs is because of the time it is taking and limited resources I have.

I am getting better results with Random Forest, So I will not be using this model for my test predictions


# Bottom Category With Random Forest

In [ ]:
# libraries
import io
import os
from datetime import datetime
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from sklearn import metrics

# Matplotlib configuration
font = { 'family': 'DejaVu Sans', 'weight': 'bold', 'size': 16 }
plt.rc('font', **font)

In [ ]:
# set random seed
np.random.seed(seed=42)

##Preprocessing

This cleaning and preprocessing is from https://github.com/rkj43/Jigsaw_Toxic_Comment_Classification/blob/main/Jigsaw_Classification_CA683.ipynb.
(That's my Github. I have used the same preprocessing here. It is a modified version of the preprocessing in this notebook https://www.kaggle.com/code/floevan/dl-and-ml-for-nlp/notebook#SIMPLE-RNN)

In [ ]:
from nltk.corpus import stopwords
import nltk
import re
import string, collections
from nltk.stem.snowball import EnglishStemmer
from nltk.stem import WordNetLemmatizer
nltk.download("stopwords")
stop_words = set(stopwords.words('english'))
tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
stemmer = EnglishStemmer()
lemmatizer = WordNetLemmatizer()

def clean_text(text):
    text = text.lower() #make text lowercase and fill na
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('\\n', '',str(text))
    text = re.sub("\[\[User.*",'',str(text))
    text = re.sub("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}",'',str(text))
    text = re.sub('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', '', text) #remove hyperlinks
    text = re.sub(r'\:(.*?)\:', '', text) #remove emoticones
    text = re.sub(r'[\w\.-]+@[\w\.-]+', '', str(text)) #remove email
    text = re.sub(r'(?<=@)\w+', '', text) #remove @
    text = re.sub(r'[0-9]+', '', text) #remove numbers
    text = re.sub("[^A-Za-z0-9 ]", '', text) #remove non alphanumeric like ['@', '#', '.', '(', ')']
    text = re.sub(r'[!"\$%&\'()*+,\-.\/:;=#@?\[\\\]^_`{|}~]*', '', text) #remove punctuations from sentences
    text = re.sub('<.*?>+', '', str(text))
    text = re.sub('[%s]' % re.escape(string.punctuation), '', str(text))
    text = re.sub('\w*\d\w*', '', str(text))
    text = tokenizer.tokenize(text)
    text = [word for word in text if not word in stop_words]
    #text = [lemmatizer.lemmatize(word) for word in text]
    text = [stemmer.stem(word) for word in text]
    final_text = ' '.join( [w for w in text if len(w)>1] ) #remove word with one letter
    return final_text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
data_train['description']=data_train['description'].fillna('')

In [ ]:
data_test['description']=data_test['description'].fillna('')

In [ ]:
from sklearn.model_selection import train_test_split

df_training, df_validation = train_test_split(
    data_train,
    test_size=0.2,
    stratify=data_train['bottom_category_text'],
    random_state=42)

In [ ]:
df_training['description']=df_training['description'].fillna('')
df_validation['description']=df_validation['description'].fillna('')


In [ ]:
df_training['title']=df_training['title'].fillna('')
df_validation['title']=df_validation['title'].fillna('')


In [ ]:
df_training['description'] = df_training['description'].apply(lambda x : clean_text(x))
df_training['title'] = df_training['title'].apply(lambda x : clean_text(x))

In [ ]:
df_validation['description'] = df_validation['description'].apply(lambda x : clean_text(x))
df_validation['title'] = df_validation['title'].apply(lambda x : clean_text(x))

In [ ]:
f"Number of products for training: {len(df_training):,}"

'Number of products for training: 196,388'

In [ ]:
f"Number of products for validation: {len(df_validation):,}"

'Number of products for validation: 49,097'

## Training

- processes the text using `CountVectorizer`, that builds a dictionary of features and transforms documents to feature vectors
- extracts term frequencies and inverse document frequencies using `TfidfTransformer`, (see [TF-IDF](https://en.wikipedia.org/wiki/Tf%E2%80%93idf))
- trains a Random forest classifer, which gives us a nice baseline

This is similar to the one david provided but I have made some changes to run things parallely and also I am using random forest instead of native bayes.
Links I have referred for the codes :
- https://scikit-learn.org/stable/modules/tree.html
- https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
- https://stackoverflow.com/questions/60080806/how-to-run-2-pipelines-in-parallel-in-scikit-learn-or-neuraxle
- https://www.kaggle.com/questions-and-answers/160593

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(verbose=100,n_jobs=-1,max_depth=200)), #n_jobs = -1 means use all cores, verbose is for logging , max_Depth to use lower depths so that we dont overfit the model
])

In [ ]:
train_text = df_training['description'] + df_training['title'] + df_training['tags']

In [ ]:
train_text = train_text.fillna('').values

In [ ]:
train_labels = df_training['bottom_category_text'].values

In [ ]:
X = train_text
Y = train_labels

text_clf = text_clf.fit(X, Y)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 12 concurrent workers.
building tree 1 of 100
building tree 2 of 100
building tree 3 of 100
building tree 4 of 100building tree 5 of 100
building tree 6 of 100
building tree 7 of 100
building tree 8 of 100
building tree 9 of 100

building tree 10 of 100
building tree 11 of 100building tree 12 of 100

building tree 13 of 100[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:   30.1s

building tree 14 of 100[Parallel(n_jobs=-1)]: Done   2 tasks      | elapsed:   32.8s

building tree 15 of 100[Parallel(n_jobs=-1)]: Done   3 tasks      | elapsed:   32.9s

building tree 16 of 100[Parallel(n_jobs=-1)]: Done   4 tasks      | elapsed:   33.6s

building tree 17 of 100[Parallel(n_jobs=-1)]: Done   5 tasks      | elapsed:   33.8s

building tree 18 of 100[Parallel(n_jobs=-1)]: Done   6 tasks      | elapsed:   34.6s

building tree 19 of 100[Parallel(n_jobs=-1)]: Done   7 tasks      | elapsed:   34.7s

building tree 20 of 100[Parallel(n_jo

##Results

In [ ]:
val_text = df_validation['description'] + df_validation['title'] + df_validation['tags']
val_text = val_text.fillna('').values
val_labels = df_validation['bottom_category_text'].values

In [ ]:
predictions = text_clf.predict(val_text)

[Parallel(n_jobs=12)]: Using backend ThreadingBackend with 12 concurrent workers.
[Parallel(n_jobs=12)]: Done   1 tasks      | elapsed:    2.3s
[Parallel(n_jobs=12)]: Done   2 tasks      | elapsed:    2.5s
[Parallel(n_jobs=12)]: Done   3 tasks      | elapsed:    2.7s
[Parallel(n_jobs=12)]: Done   4 tasks      | elapsed:    2.9s
[Parallel(n_jobs=12)]: Done   5 tasks      | elapsed:    3.1s
[Parallel(n_jobs=12)]: Done   6 tasks      | elapsed:    3.3s
[Parallel(n_jobs=12)]: Done   7 tasks      | elapsed:    3.5s
[Parallel(n_jobs=12)]: Done   8 tasks      | elapsed:    3.7s
[Parallel(n_jobs=12)]: Done   9 tasks      | elapsed:    3.8s
[Parallel(n_jobs=12)]: Done  10 tasks      | elapsed:    4.0s
[Parallel(n_jobs=12)]: Done  11 tasks      | elapsed:    4.2s
[Parallel(n_jobs=12)]: Done  12 tasks      | elapsed:    4.4s
[Parallel(n_jobs=12)]: Done  13 tasks      | elapsed:    4.6s
[Parallel(n_jobs=12)]: Done  14 tasks      | elapsed:    4.8s
[Parallel(n_jobs=12)]: Done  15 tasks      | elaps

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(val_labels,predictions))

In [ ]:
f1 = metrics.f1_score(val_labels, predictions, average='weighted')

In [ ]:

f"F1 score: {f1:.4f}"

'F1 score: 0.4519'

In [ ]:
metrics.confusion_matrix(val_labels, predictions)

array([[ 2,  0,  0, ...,  0,  0,  0],
       [ 0,  2,  0, ...,  0,  0,  0],
       [ 0,  0,  7, ...,  0,  0,  0],
       ...,
       [ 0,  0,  0, ..., 13,  0,  1],
       [ 0,  0,  0, ...,  0, 13,  0],
       [ 0,  0,  0, ...,  0,  0, 12]])

We need to do the same preprocessing we do for training on test data too.

In [ ]:
data_test['description']=data_test['description'].fillna('')
data_test['title']=data_test['title'].fillna('')
data_test['description'] = data_test['description'].apply(lambda x : clean_text(x))
data_test['title'] = data_test['title'].apply(lambda x : clean_text(x))

In [ ]:
test_text = data_test['description'] + data_test['title'] + data_test['tags']
test_text = test_text.fillna('').values


In [ ]:
predictions_test = text_clf.predict(test_text)

[Parallel(n_jobs=12)]: Using backend ThreadingBackend with 12 concurrent workers.
[Parallel(n_jobs=12)]: Done   1 tasks      | elapsed:    1.3s
[Parallel(n_jobs=12)]: Done   2 tasks      | elapsed:    1.4s
[Parallel(n_jobs=12)]: Done   3 tasks      | elapsed:    1.5s
[Parallel(n_jobs=12)]: Done   4 tasks      | elapsed:    1.6s
[Parallel(n_jobs=12)]: Done   5 tasks      | elapsed:    1.7s
[Parallel(n_jobs=12)]: Done   6 tasks      | elapsed:    1.8s
[Parallel(n_jobs=12)]: Done   7 tasks      | elapsed:    1.9s
[Parallel(n_jobs=12)]: Done   8 tasks      | elapsed:    2.0s
[Parallel(n_jobs=12)]: Done   9 tasks      | elapsed:    2.1s
[Parallel(n_jobs=12)]: Done  10 tasks      | elapsed:    2.2s
[Parallel(n_jobs=12)]: Done  11 tasks      | elapsed:    2.3s
[Parallel(n_jobs=12)]: Done  12 tasks      | elapsed:    2.4s
[Parallel(n_jobs=12)]: Done  13 tasks      | elapsed:    2.5s
[Parallel(n_jobs=12)]: Done  14 tasks      | elapsed:    2.7s
[Parallel(n_jobs=12)]: Done  15 tasks      | elaps

Mapping to bottom_category_id

In [ ]:
map1 = set(data_train[['bottom_category_id','bottom_category_text']].to_records(index=False).tolist())


In [ ]:
mapping = {y:x for (x,y) in map1}

In [ ]:
test_predictions_mapped = [mapping[x] for x in predictions_test]

Since I am getting better F1 for bottom category with Random forest as compared to BERT. I will be using this model for my final predictions for submission

In [ ]:
pd.DataFrame(test_predictions_mapped).to_csv('/content/gdrive/MyDrive/bottomcategorypredictions.csv')

# Top Category With Random Forest

In [ ]:
# libraries
import io
import os
from datetime import datetime
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from sklearn import metrics

# Matplotlib configuration
font = { 'family': 'DejaVu Sans', 'weight': 'bold', 'size': 16 }
plt.rc('font', **font)

In [ ]:
# set random seed
np.random.seed(seed=42)

##Preprocessing

This cleaning and preprocessing is from https://github.com/rkj43/Jigsaw_Toxic_Comment_Classification/blob/main/Jigsaw_Classification_CA683.ipynb.
(That's my Github. I have used the same preprocessing here. It is a modified version of the preprocessing in this notebook https://www.kaggle.com/code/floevan/dl-and-ml-for-nlp/notebook#SIMPLE-RNN)

In [ ]:
from nltk.corpus import stopwords
import nltk
import re
import string, collections
from nltk.stem.snowball import EnglishStemmer
from nltk.stem import WordNetLemmatizer
nltk.download("stopwords")
stop_words = set(stopwords.words('english'))
tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
stemmer = EnglishStemmer()
lemmatizer = WordNetLemmatizer()

def clean_text(text):
    text = text.lower() #make text lowercase and fill na
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('\\n', '',str(text))
    text = re.sub("\[\[User.*",'',str(text))
    text = re.sub("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}",'',str(text))
    text = re.sub('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', '', text) #remove hyperlinks
    text = re.sub(r'\:(.*?)\:', '', text) #remove emoticones
    text = re.sub(r'[\w\.-]+@[\w\.-]+', '', str(text)) #remove email
    text = re.sub(r'(?<=@)\w+', '', text) #remove @
    text = re.sub(r'[0-9]+', '', text) #remove numbers
    text = re.sub("[^A-Za-z0-9 ]", '', text) #remove non alphanumeric like ['@', '#', '.', '(', ')']
    text = re.sub(r'[!"\$%&\'()*+,\-.\/:;=#@?\[\\\]^_`{|}~]*', '', text) #remove punctuations from sentences
    text = re.sub('<.*?>+', '', str(text))
    text = re.sub('[%s]' % re.escape(string.punctuation), '', str(text))
    text = re.sub('\w*\d\w*', '', str(text))
    text = tokenizer.tokenize(text)
    text = [word for word in text if not word in stop_words]
    #text = [lemmatizer.lemmatize(word) for word in text]
    text = [stemmer.stem(word) for word in text]
    final_text = ' '.join( [w for w in text if len(w)>1] ) #remove word with one letter
    return final_text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
data_train['description']=data_train['description'].fillna('')

In [ ]:
data_test['description']=data_test['description'].fillna('')

In [ ]:
from sklearn.model_selection import train_test_split

df_training, df_validation = train_test_split(
    data_train,
    test_size=0.2,
    stratify=data_train['top_category_text'],
    random_state=42)

In [ ]:
df_training['description']=df_training['description'].fillna('')
df_validation['description']=df_validation['description'].fillna('')


In [ ]:
df_training['title']=df_training['title'].fillna('')
df_validation['title']=df_validation['title'].fillna('')


In [ ]:
# df_training['description'] = df_training['description'].apply(lambda x : clean_text(x))
# df_training['title'] = df_training['title'].apply(lambda x : clean_text(x))

In [ ]:
# df_validation['description'] = df_validation['description'].apply(lambda x : clean_text(x))
# df_validation['title'] = df_validation['title'].apply(lambda x : clean_text(x))

In [ ]:
f"Number of products for training: {len(df_training):,}"

'Number of products for training: 196,388'

In [ ]:
f"Number of products for validation: {len(df_validation):,}"

'Number of products for validation: 49,097'

## Training

- processes the text using `CountVectorizer`, that builds a dictionary of features and transforms documents to feature vectors
- extracts term frequencies and inverse document frequencies using `TfidfTransformer`, (see [TF-IDF](https://en.wikipedia.org/wiki/Tf%E2%80%93idf))
- trains a Random forest classifer, which gives us a nice baseline

This is similar to the one david provided but I have made some changes to run things parallely and also I am using random forest instead of native bayes.
Links I have referred for the codes :
- https://scikit-learn.org/stable/modules/tree.html
- https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
- https://stackoverflow.com/questions/60080806/how-to-run-2-pipelines-in-parallel-in-scikit-learn-or-neuraxle
- https://www.kaggle.com/questions-and-answers/160593

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(verbose=100,n_jobs=-1,max_depth=200)), #n_jobs = -1 means use all cores, verbose is for logging , max_Depth to use lower depths so that we dont overfit the model
])

In [ ]:
train_text = df_training['description'] + df_training['title'] + df_training['tags']

In [ ]:
train_text = train_text.fillna('').values

In [ ]:
train_labels = df_training['top_category_text'].values

In [ ]:
X = train_text
Y = train_labels

text_clf = text_clf.fit(X, Y)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 12 concurrent workers.
building tree 1 of 100
building tree 2 of 100
building tree 3 of 100
building tree 4 of 100
building tree 5 of 100building tree 6 of 100
building tree 7 of 100

building tree 8 of 100building tree 9 of 100
building tree 10 of 100
building tree 11 of 100
building tree 12 of 100

building tree 13 of 100[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:   36.3s

building tree 14 of 100[Parallel(n_jobs=-1)]: Done   2 tasks      | elapsed:   37.1s

building tree 15 of 100[Parallel(n_jobs=-1)]: Done   3 tasks      | elapsed:   37.8s

building tree 16 of 100[Parallel(n_jobs=-1)]: Done   4 tasks      | elapsed:   38.0s

building tree 17 of 100[Parallel(n_jobs=-1)]: Done   5 tasks      | elapsed:   38.2s

building tree 18 of 100[Parallel(n_jobs=-1)]: Done   6 tasks      | elapsed:   39.6s

building tree 19 of 100[Parallel(n_jobs=-1)]: Done   7 tasks      | elapsed:   39.6s

building tree 20 of 100[Parallel(n_jo

##Results

In [ ]:
val_text = df_validation['description'] + df_validation['title'] + df_validation['tags']
val_text = val_text.fillna('').values
val_labels = df_validation['top_category_text'].values

In [ ]:
predictions = text_clf.predict(val_text)

[Parallel(n_jobs=12)]: Using backend ThreadingBackend with 12 concurrent workers.
[Parallel(n_jobs=12)]: Done   1 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   2 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   3 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   4 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   5 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   6 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   7 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   8 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   9 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  10 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  11 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  12 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  13 tasks      | elapsed:    0.2s
[Parallel(n_jobs=12)]: Done  14 tasks      | elapsed:    0.2s
[Parallel(n_jobs=12)]: Done  15 tasks      | elaps

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(val_labels,predictions))

In [ ]:
f1 = metrics.f1_score(val_labels, predictions, average='weighted')

In [ ]:

f"F1 score: {f1:.4f}"

'F1 score: 0.7057'

In [ ]:
metrics.confusion_matrix(val_labels, predictions)

array([[ 1180,     0,     1,    17,     1,   355,   163,     3,   677,
           32,     3,     6,     5,     6,    73],
       [    2,   450,     0,     0,     5,    32,   167,     2,  1014,
            0,    33,     8,     2,    29,     4],
       [    4,     0,   632,     3,     0,    58,    62,     2,   550,
            5,     4,     0,     2,     2,     5],
       [    7,     0,     7,  1438,     0,    91,   109,     0,   809,
            5,     9,     2,     0,     7,     1],
       [    1,     0,     3,     0,   864,    12,   106,     3,   436,
            4,    33,     2,     1,     9,     2],
       [   11,     0,     1,     3,     0,  5288,    14,     0,  1070,
            0,     1,     2,     5,     2,     9],
       [   12,    23,     3,    16,    14,    37,  6183,     6,  2069,
           27,    39,     4,    17,    13,     7],
       [    1,     0,     8,     2,     2,     5,    99,   929,   565,
            1,    14,     1,     0,     8,     0],
       [    2,    14,   

Since I am **not** getting better F1 for bottom category with Random forest as compared to BERT. I will **not**  be using this model for my final predictions for submission.
Will stick to BERT

# Color With Random Forest

In [ ]:
# libraries
import io
import os
from datetime import datetime
import pandas as pd
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from sklearn import metrics

# Matplotlib configuration
font = { 'family': 'DejaVu Sans', 'weight': 'bold', 'size': 16 }
plt.rc('font', **font)

In [ ]:
# set random seed
np.random.seed(seed=42)

##Preprocessing

This cleaning and preprocessing is from https://github.com/rkj43/Jigsaw_Toxic_Comment_Classification/blob/main/Jigsaw_Classification_CA683.ipynb.
(That's my Github. I have used the same preprocessing here. It is a modified version of the preprocessing in this notebook https://www.kaggle.com/code/floevan/dl-and-ml-for-nlp/notebook#SIMPLE-RNN)

In [ ]:
from nltk.corpus import stopwords
import nltk
import re
import string, collections
from nltk.stem.snowball import EnglishStemmer
from nltk.stem import WordNetLemmatizer
nltk.download("stopwords")
stop_words = set(stopwords.words('english'))
tokenizer = nltk.tokenize.RegexpTokenizer(r'\w+')
stemmer = EnglishStemmer()
lemmatizer = WordNetLemmatizer()

def clean_text(text):
    text = text.lower() #make text lowercase and fill na
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('\\n', '',str(text))
    text = re.sub("\[\[User.*",'',str(text))
    text = re.sub("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}",'',str(text))
    text = re.sub('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', '', text) #remove hyperlinks
    text = re.sub(r'\:(.*?)\:', '', text) #remove emoticones
    text = re.sub(r'[\w\.-]+@[\w\.-]+', '', str(text)) #remove email
    text = re.sub(r'(?<=@)\w+', '', text) #remove @
    text = re.sub(r'[0-9]+', '', text) #remove numbers
    text = re.sub("[^A-Za-z0-9 ]", '', text) #remove non alphanumeric like ['@', '#', '.', '(', ')']
    text = re.sub(r'[!"\$%&\'()*+,\-.\/:;=#@?\[\\\]^_`{|}~]*', '', text) #remove punctuations from sentences
    text = re.sub('<.*?>+', '', str(text))
    text = re.sub('[%s]' % re.escape(string.punctuation), '', str(text))
    text = re.sub('\w*\d\w*', '', str(text))
    text = tokenizer.tokenize(text)
    text = [word for word in text if not word in stop_words]
    #text = [lemmatizer.lemmatize(word) for word in text]
    text = [stemmer.stem(word) for word in text]
    final_text = ' '.join( [w for w in text if len(w)>1] ) #remove word with one letter
    return final_text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Color ID 8 or Color Grey has only one value/row in the training dataset. Therefore I am dropping it

In [ ]:
data_train = data_train[data_train['color_id'] != 8]

In [ ]:
data_train['description']=data_train['description'].fillna('')

In [ ]:
data_test['description']=data_test['description'].fillna('')

In [ ]:
from sklearn.model_selection import train_test_split

df_training, df_validation = train_test_split(
    data_train,
    test_size=0.2,
    stratify=data_train['color_text'],
    random_state=42)

In [ ]:
df_training['description']=df_training['description'].fillna('')
df_validation['description']=df_validation['description'].fillna('')


In [ ]:
df_training['title']=df_training['title'].fillna('')
df_validation['title']=df_validation['title'].fillna('')


In [ ]:
# df_training['description'] = df_training['description'].apply(lambda x : clean_text(x))
# df_training['title'] = df_training['title'].apply(lambda x : clean_text(x))

In [ ]:
# df_validation['description'] = df_validation['description'].apply(lambda x : clean_text(x))
# df_validation['title'] = df_validation['title'].apply(lambda x : clean_text(x))

In [ ]:
f"Number of products for training: {len(df_training):,}"

'Number of products for training: 196,387'

In [ ]:
f"Number of products for validation: {len(df_validation):,}"

'Number of products for validation: 49,097'

## Training

- processes the text using `CountVectorizer`, that builds a dictionary of features and transforms documents to feature vectors
- extracts term frequencies and inverse document frequencies using `TfidfTransformer`, (see [TF-IDF](https://en.wikipedia.org/wiki/Tf%E2%80%93idf))
- trains a Random forest classifer, which gives us a nice baseline

This is similar to the one david provided but I have made some changes to run things parallely and also I am using random forest instead of native bayes.
Links I have referred for the codes :
- https://scikit-learn.org/stable/modules/tree.html
- https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
- https://stackoverflow.com/questions/60080806/how-to-run-2-pipelines-in-parallel-in-scikit-learn-or-neuraxle
- https://www.kaggle.com/questions-and-answers/160593

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', RandomForestClassifier(verbose=100,n_jobs=-1,max_depth=200)),#n_jobs = -1 means use all cores, verbose is for logging , max_Depth to use lower depths so that we dont overfit the model
])

In [ ]:
train_text = df_training['description'] + df_training['title'] + df_training['tags']

In [ ]:
train_text = train_text.fillna('').values

In [ ]:
train_labels = df_training['color_text'].values

In [ ]:
X = train_text
Y = train_labels

text_clf = text_clf.fit(X, Y)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 12 concurrent workers.
building tree 1 of 100
building tree 2 of 100
building tree 3 of 100
building tree 4 of 100
building tree 5 of 100
building tree 6 of 100building tree 7 of 100
building tree 8 of 100
building tree 9 of 100

building tree 10 of 100
building tree 11 of 100
building tree 12 of 100
building tree 13 of 100[Parallel(n_jobs=-1)]: Done   1 tasks      | elapsed:  1.0min

building tree 14 of 100[Parallel(n_jobs=-1)]: Done   2 tasks      | elapsed:  1.1min

building tree 15 of 100[Parallel(n_jobs=-1)]: Done   3 tasks      | elapsed:  1.1min

building tree 16 of 100[Parallel(n_jobs=-1)]: Done   4 tasks      | elapsed:  1.1min

building tree 17 of 100[Parallel(n_jobs=-1)]: Done   5 tasks      | elapsed:  1.1min

building tree 18 of 100[Parallel(n_jobs=-1)]: Done   6 tasks      | elapsed:  1.1min

building tree 19 of 100[Parallel(n_jobs=-1)]: Done   7 tasks      | elapsed:  1.1min

building tree 20 of 100[Parallel(n_jo

##Results

In [ ]:
val_text = df_validation['description'] + df_validation['title'] + df_validation['tags']
val_text = val_text.fillna('').values
val_labels = df_validation['color_text'].values

In [ ]:
predictions = text_clf.predict(val_text)

[Parallel(n_jobs=12)]: Using backend ThreadingBackend with 12 concurrent workers.
[Parallel(n_jobs=12)]: Done   1 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   2 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   3 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   4 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   5 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   6 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   7 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   8 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done   9 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  10 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  11 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  12 tasks      | elapsed:    0.1s
[Parallel(n_jobs=12)]: Done  13 tasks      | elapsed:    0.2s
[Parallel(n_jobs=12)]: Done  14 tasks      | elapsed:    0.3s
[Parallel(n_jobs=12)]: Done  15 tasks      | elaps

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(val_labels,predictions))

              precision    recall  f1-score   support

       beige       0.62      0.17      0.27      2533
       black       0.30      0.74      0.42      6508
        blue       0.42      0.44      0.43      5395
      bronze       0.73      0.25      0.38       584
       brown       0.51      0.55      0.53      4578
       clear       0.76      0.44      0.56      1267
      copper       0.75      0.33      0.46       447
        gold       0.62      0.41      0.49      2146
       green       0.52      0.27      0.36      3135
      orange       0.43      0.12      0.19      1069
       other       0.52      0.18      0.27      1984
        pink       0.49      0.30      0.37      3100
      purple       0.45      0.16      0.24      1328
     rainbow       0.63      0.20      0.31      1045
         red       0.51      0.26      0.35      3049
   rose gold       0.37      0.09      0.14       231
      silver       0.61      0.62      0.61      2978
       white       0.38    

In [ ]:
f1 = metrics.f1_score(val_labels, predictions, average='weighted')

In [ ]:

f"F1 score: {f1:.4f}"

'F1 score: 0.4070'

In [ ]:
metrics.confusion_matrix(val_labels, predictions)

array([[ 438,  776,  207,    3,  402,    4,    3,   10,   42,    8,   22,
          43,    7,    1,   43,    1,   23,  491,    9],
       [  24, 4794,  320,    5,  348,   11,    3,   42,   72,   13,   37,
          66,   10,   10,   65,    3,  111,  559,   15],
       [  26, 1390, 2366,    4,  172,   21,    5,   27,  103,   19,   52,
         159,   54,   23,  113,    2,  116,  708,   35],
       [   4,  128,   17,  148,   86,    2,   10,   59,    3,    2,    1,
           6,    1,    2,    7,    4,   69,   33,    2],
       [  42, 1121,  193,    4, 2523,   13,    3,   27,   55,   12,   33,
          33,   14,   10,   51,    0,   66,  361,   17],
       [   5,  263,   48,    0,   80,  559,    0,   16,   20,    2,    2,
          12,    3,    1,    8,    1,   50,  197,    0],
       [   0,   87,   20,    4,   57,    0,  149,   13,    4,    0,    2,
           4,    0,    1,    4,    1,   62,   38,    1],
       [  11,  474,   82,   13,   95,    8,    6,  878,   33,    1,    3,
         

Since I am **not** getting better F1 for color with Random forest as compared to BERT. I will **not**  be using this model for my final predictions for submission.
Will stick to BERT

#Submission

In [ ]:
from google.colab import drive

drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
import pandas as pd

df_predictions = pd.read_csv('/content/gdrive/MyDrive/prediction.csv')

In [ ]:
df_bottom = pd.read_csv('/content/gdrive/MyDrive/bottomcategorypredictions.csv')

In [ ]:
df_color = pd.read_csv('/content/gdrive/MyDrive/colorpredictions.csv')

In [ ]:
df_predictions

Unnamed: 0  product_id  top_category_id
0               0  1323824090                6
1               1   544175230                6
2               2   895779370                6
3               3   641102090                6
4               4  1340603160                8
...           ...         ...              ...
27114       27114   809100810                4
27115       27115  1358895140                4
27116       27116  1366768240                4
27117       27117  1168372040                4
27118       27118  1157860460                4

[27119 rows x 3 columns]

In [ ]:
df_predictions['bottom_category_id'] = df_bottom['0']
df_predictions['color_id'] = df_color['0']

In [ ]:
df_predictions

Unnamed: 0  product_id  top_category_id  bottom_category_id  color_id
0               0  1323824090                6                6883        10
1               1   544175230                6                6883         9
2               2   895779370                6                6886         9
3               3   641102090                6                6886         9
4               4  1340603160                8                6883        14
...           ...         ...              ...                 ...       ...
27114       27114   809100810                4                6359         4
27115       27115  1358895140                4                2643         4
27116       27116  1366768240                4                2012         2
27117       27117  1168372040                4                 324         4
27118       27118  1157860460                4                 347         4

[27119 rows x 5 columns]

In [ ]:
student_id = "22262108"
df_predictions[
    ["product_id", "top_category_id", "bottom_category_id", "color_id"]
].to_parquet(f"predictions_{student_id}.parquet")

In [ ]:
%cp -av /content/predictions_22262108.parquet /content/gdrive/MyDrive/
#code to move the the file  to google drive

'/content/predictions_22262108.parquet' -> '/content/gdrive/MyDrive/predictions_22262108.parquet'


#What can I do to improve this?

Here's what I think can be improved
*   Investigate more and tune BERT to get better results for bottom category
*   For the models, I can try different transformers that are supported by the multimodal package to see if I get better results
*   Writing Global Functions. I need to use functions so I can just set variables. That will easily reduce the repeating code and make the notebook better.
* Try hyperparameter tuning for Random Forest.
* Try XGboost
* Your feedback is crucial and important for me to learn and improve my skills.

Thank you for going through this notebook. I hope I was atleast able to give something different with my approaches
